src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
             => (Header -> ConduitM ByteString o m ())
             -> ConduitM TarChunk o m ()
 withEntries = peekForever . withEntry
+
+
+-- | Extract a tarball, similarly to `withEntries`, but instead of dealing directly with tar format,
+-- this conduit allows you to work directly on file abstractions `FileInfo`. For now support is
+-- minimal:
+--
+-- * Old v7 tar format.
+-- * ustar: POSIX 1003.1-1988 format
+-- * and only portions of GNU format:
+--   * 'L' type - long file names, but only up to 4096 chars to prevent DoS attack
+--   * other types are simply discarded
+--
+-- /Note/ - Here is a really good reference for specifics of different tar formats:
+-- <https://github.com/libarchive/libarchive/wiki/ManPageTar5>
+withFileInfo :: MonadThrow m
+             => (FileInfo -> ConduitM ByteString o m ())
+             -> ConduitM TarChunk o m ()
+withFileInfo inner = do
+    mc <- await
+    case mc of
+        Nothing -> return ()
+        Just (ChunkHeader h)
+            | headerLinkIndicator h >= 55 -> do
+                when (headerMagicVersion h == gnuTarMagicVersion) $ handleGnuTarHeader h
+                withFileInfo inner
```

Comment:
Minor optimization point. I believe the worker/wrapper transformation could be applied here to avoid potentially building up stack frames. This would look like:

```haskell
withFileInfo inner = go
  where
    go = do
      mc <- await
      ...
        s/withFileInfo inner/go
```

Reply:
Done